k-distribution/src/main/scripts/lib/pyk/kastManip.py

```diff
@@ -165,12 +193,94 @@ def addOccurance(k):
     collectBottomUp(kast, addOccurance)
     return occurances
 
+def countVarOccurances(kast, numOccurances = None):
+    """Count the number of occurances of each variable in a proof.
+
+    -   Input: Kast term.
+    -   Output: Map of variable names to their number of occurances.
+    """
+    numOccurances = {} if numOccurances is None else numOccurances
+    def _getNumOccurances(_kast):
+        if isKVariable(_kast):
+            vName = _kast['name']
+            if vName in numOccurances:
+                numOccurances[vName] += 1
+            else:
+                numOccurances[vName] = 1
+    collectBottomUp(kast, _getNumOccurances)
+    return numOccurances
+
+def collectFreeVars(kast):
+    return list(countVarOccurances(kast).keys())
+
 def flattenLabel(label, kast):
-    if isKApply(kast) and kast["label"] == label:
-        constraints = [ flattenLabel(label, arg) for arg in kast["args"] ]
-        return [ c for cs in constraints for c in cs ]
+    """Given a binary tree of a label, return all the leaves.
+
+    -   Input: label of binary tree, and kast term.
+    -   Output: list of leaves of binary tree (singleton list for no occurance of label at top).
+    """
+    if isKApply(kast) and kast['label'] == label:
+        items = [ flattenLabel(label, arg) for arg in kast['args'] ]
+        return [ c for cs in items for c in cs ]
     return [kast]
 
+def splitConfigAndConstraints(kast):
+    """Split the configuration/term from the constraints.
+
+    -   Input: kast conjunct representing a constrained term.
+    -   Output: tuple of term and constraint.
+    """
+    conjuncts = flattenLabel('#And', kast)
+    term = None
+    constraints = []
+    for c in conjuncts:
+        if isKApply(c) and isCellKLabel(c['label']):
+            term = c
+        else:
+            constraints.append(c)
+    constraint = buildAssoc(KConstant('#Top'), '#And', constraints)
+    return (term, constraint)
+
+def findCommonItems(l1, l2):
+    common = []
+    for i in l1:
+        if i in l2:
+            common.append(i)
+    newL1 = []
+    newL2 = []
+    for i in l1:
+        if not i in common:
+            newL1.append(i)
+    for i in l2:
+        if not i in common:
+            newL2.append(i)
+    return (common, newL1, newL2)
+
+def propogateUpConstraints(k):
+    """Try to propogate common constraints up disjuncts.
+
+    -   Input: kast disjunct of constrained terms (conjuncts).
+    -   Output: kast where common constraints in the disjunct have been propogated up.
```

Comment:
`propagated` Typo? Multiple locations, function names and comments.

Reply:
Fixed.